torch/_subclasses/meta_utils.py

```diff
@@ -395,11 +395,32 @@ def is_c_of_r(complex_dtype, real_dtype):
                 else:
                     is_leaf = safe_is_leaf(t)
                     sizes, strides, storage_offset = sym_sizes_strides_storage_offset(t)
-                    r = callback(
-                        lambda: torch.empty_strided(
-                            sizes, strides, dtype=t.dtype, device="meta"
+
+                    # If we have a subclass that desugars into dense tensors,
+                    # perform our callback on each inner tensor.
+                    if is_traceable_wrapper_subclass(t):
+
+                        def empty_create(inner_t):
+                            is_leaf = safe_is_leaf(inner_t)
+                            (
+                                sizes,
+                                strides,
+                                storage_offset,
+                            ) = sym_sizes_strides_storage_offset(inner_t)
```

Comment:
This looks wrong, you will attribute this the Source of the outer subclass object, but you actually need a refined source that can reference the inner part of the subclass tensor.

E.g., let's say you have DTensor.data, then the source here must be `dtensor.data` not `dtensor`.

This also implies we may need to change the flatten API, because not only do you need to know how to flatten tensors, you also need to know how to project them out of the tensor, which may be done most conveniently with a source assignment to each field. Or I suppose you can do it as is by flattening and then numeric indexing, which is worse for perf but I guess it will be OK if we CSE the flattening process.

Reply:
darn, good point.


Hmm, updating the contract for `__tensor_flatten__` feels better (so any user-facing logs on source are more readable).


> Or I suppose you can do it as is by flattening and then numeric indexing, which is worse for perf but I guess it will be OK if we CSE the flattening process.

Quick question to confirm my understanding: you're implying that how we choose to represent source here has runtime implications. Are you talking about guard printing + guard evaluation? Or something else. I'm imagining that when we eventually spit out symbolic shape guards about the sizes of our inner tensor, then if we do the numeric indexing approach you're talking about, the guards would look something like:

```
assert double_tensor_input.__tensor_flatten__()[0][1].shape[0] > 2
```

Instead of this (which is clearly cheaper, but you pointed out that the above could be made cheaper with CSE)
```
assert double_tensor_input.b.shape[0] > 2
```

Reply:
Yes, the former wouldn't require a contract change on tensor flatten.